src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,306 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <chrono>
+
+#include "absl/container/btree_map.h"
+#include "absl/strings/str_format.h"
+#include "ray/common/grpc_util.h"
+#include "ray/rpc/client_call.h"
+#include "ray/rpc/grpc_client.h"
+
+namespace ray {
+namespace rpc {
+
+#define INVOKE_RETRYABLE_RPC_CALL(retryable_rpc_client,                       \
+                                  SERVICE,                                    \
+                                  METHOD,                                     \
+                                  request,                                    \
+                                  callback,                                   \
+                                  rpc_client,                                 \
+                                  method_timeout_ms)                          \
+  (retryable_rpc_client->CallMethod<SERVICE, METHOD##Request, METHOD##Reply>( \
+      &SERVICE::Stub::PrepareAsync##METHOD,                                   \
+      rpc_client,                                                             \
+      #SERVICE ".grpc_client." #METHOD,                                       \
+      request,                                                                \
+      callback,                                                               \
+      method_timeout_ms))
+
+/// \class Executor
+/// Executor saves operation and support retries.
+class Executor {
+ public:
+  Executor(std::function<void(const ray::Status &)> abort_callback)
+      : abort_callback_(std::move(abort_callback)) {}
+
+  /// This function is used to execute the given operation.
+  ///
+  /// \param operation The operation to be executed.
+  void Execute(std::function<void()> operation) {
+    operation_ = std::move(operation);
+    operation_();
+  }
+
+  /// This function is used to retry the given operation.
+  void Retry() { operation_(); }
+
+  void Abort(const ray::Status &status) { abort_callback_(status); }
+
+ private:
+  std::function<void(ray::Status)> abort_callback_;
+  std::function<void()> operation_;
+};
+
+class RetryableGrpcClient : public std::enable_shared_from_this<RetryableGrpcClient> {
```

Comment:
I'm thinking if we could separate the logic of the retry vs logic of sending out 1 rpc. That is, let a RetryableGrpcClient to hold a GrpcClient. So that the latter focuses on sending 1 rpc, and the former focuses on the retry. I am sketching some code to see if this works.

Reply:
I am positive since GcsRpcClient is already a retry layer on top of GrpcClient.